### ISSUE TITLE
Evaluate the possibility of have 0 deps with JSP runtime from Jetty itself

### ISSUE THREAD

**Issue by ludoch** [GoogleCloudPlatform/appengine-java-vm-runtime#66]
Ideally, the JSP runtime should be installable only in the WEB-INF user application classpath, when the JSPs are precompiled.

**Comment by gregw** [comment]
Just to double check we are on the same page, the desire is to deploy a war file with precompile JSPs to a server that does not have JSP enabled and for it to work fine.      The intention being to have the JSP library version independent of the container version and thus such a precompiled war would be able to be redeployed on an updated container without be re-precompiled.

I think we need a name for this... embedded JSP? Shaded JSP? Internalized JSP?

**Comment by ludoch** [comment]
Yes, this is the intent.
I think there are a few classes that might currently be needed from the
Jetty class loader::

  org/apache/jasper/runtime/HttpJspBase.class
  org/apache/jasper/runtime/InstanceManagerFactory.class
  org/apache/jasper/runtime/JspSourceDependent.class

maybe more (or less)...

Not a top priority, but I think it was achievable with Jetty6 IIRC.

**Comment by jboynes** [comment]
Couple of issues come to mind. Firstly, there are updates made to the JSP runtime in response to issues found see https://goo.gl/RlC0dY for issues in Jasper just this year. If the JSP runtime is shipped with the container then the application will pick them up. This happens far more frequently than changes to the API that would require re-translation.

Secondly,  when run with a SecurityManager there are some privileged operations performed by the runtime that would not normally be granted to the application. If the runtime was packaged with the application this would be harder to enforce.

**Comment by gregw** [comment]
@jboynes Yep this approach would really be locking in JSP bugs and would require the application to be recompiled and redeployed.   There may also be issues with a baked in JSP getting in advance of the container - ie needing a container fix that is not yet deployed.

Perhaps there is a third way.    Perhaps we can somehow tag which JSP version was used to precompile an application and then the container can provide a backwards compatibility deployment mode.  Ie the container would still provide the JSP jars, but could choose to provide old ones to an old application in a compatibility mode?

In jetty, this could be done in a similar way to how we pick our ALPN implementation module based on the JVM version.  We could pick a JSP runtime module based on a JSP version obtained from the application.   So long as we start baking in the JSP version now, there would be no need to do work to support this until we see a breaking change in the private API between a precompiled JSP and the JSP runtime.

**Comment by jboynes** [comment]
@gregw wouldn't that have the same issues with the JSP version getting ahead of the container?

**Comment by jboynes** [comment]
Another option that has been proposed on the Jasper list is to perform bulk translation/compilation during initialization time to ensure everything was in sync. That would prolong startup but would avoid delaying live requests.

**Comment by janbartel** [comment]
FWIW I've tested precompiled jsps with the jsp & jstl jars inside WEB-INF/lib and it just works under jetty.

Non-precompiled jsps don't currently work and would need some fiddling about to try and make work.

**Comment by gregw** [comment]
@jboynes The JSP could not get ahead of the container as the container would still be providing the JSP runtime.  Only difference to now is that we could have it provide 2 or more different versions on the runtime, so old apps could be run after a container upgrade... hopefully with some kind of warning.

With respect to your other suggestion, the bulk translation/compilation might be tricky... but as an intermediate step we could perhaps strip out the precompilation, and leave the application in a compile on demand state, which would use the APIs that have some level of backward compatibility.

This would slow down old deployments, but give them an incentive to redeploy.

**Comment by jboynes** [comment]
@gregw I was thinking of the case where the container might provide 4.0.0 and 4.0.1 but the pre-compiled classes might require 4.0.2. If that version ever got deployed/staged there could be container images unable to run it (e.g. during a rolling upgrade or rollback scenario)

**Comment by janbartel** [comment]
Just to lob another thought into the mix ...  don't forget that in addition to the jsp and el jars, there would also be the jstl jars. So it would be combinations of jsp/el and jstl versions for which backward compatibility would need to be maintained.

**Comment by gregw** [comment]
@jboynes I guess that is not really a problem as the precompilation is done by the SDK, which should not be able to get ahead of the server.

I was thinking that if the SDK and servers supported 3.8.z say and we wanted to upgrade to 4.0.x, then our unit tests should detect any change in the API between a compiled JSP and the runtime, so we would know that there is a breaking change.     We would then create a server that could run either 3.8.z or 4.0.x and release a new SDK that built to 4.0.x.     The old applications and people who did not update their SDKs would still work in the compatibility mode (hopefully getting warnings to upgrade/redeploy).   New deploys would be 4.0.x only.

Hopefully, before the next breaking API change comes in 4.1.y, the support for 3.8.z would be EOL'd and turned off.  That way we would not need to support it forever, but would also not need to get the entire world to switch on the same instant.

**Comment by ludoch** [comment]
@janbartel 2 Questions for you:
1/ for the first comment saying its working, you do no enable the jsp module, and you bundle in web-inf/lib which exact jars (from a Jetty distro/Maven): lib/apache-jsp/*.jar and lib/apache-jstl/*.jar ?
2/ not 100% related, but still puzzled about the fact that I had to add: 
https://github.com/GoogleCloudPlatform/appengine-java-vm-runtime/blob/jetty93/appengine-jetty-managed-runtime/src/test/java/com/google/apphosting/vmruntime/jetty9/JettyRunner.java
line 143 and 144 to make the test pass... Where is it supposed to be done automatically by Jetty in non test env? Any better way of doing it?

**Comment by janbartel** [comment]
@ludoch 
1/ The test wasn't particularly sophisticated - I precompiled the jsps in jetty's standard test webapp. I threw all the jars that would normally be in the distro's lib/apache-jsp and lib/apache-jstl into the webapp's WEB-INF/lib. I made a fresh base in a standard jetty distro without jsp or jstl enabled and deployed the webapp. I can investigate more closely exactly which jars are necessary.

2/If you have disabled AnnotationConfiguration, then you won't get any of the ServletContainerInitializers found, and jsp uses one to set itself up. So basically all of the jsp setup machinery is missing - it just so happens that the lack of a SimpleInstanceManager is highly visible. Let me look into a way to reanable AnnotationConfiguration without necessitating scanning of jars (which you presumably want to avoid).

**Comment by ludoch** [comment]
Great!

**Comment by janbartel** [comment]
@ludoch 
1/ The list of jars I used is as follows (just substitute 9.3.4-SNAPSHOT for your version of jetty):
org.apache.taglibs.taglibs-standard-spec-1.2.1.jar
org.apache.taglibs.taglibs-standard-impl-1.2.1.jar  
org.mortbay.jasper.apache-jsp-8.0.23.M1.jar
org.mortbay.jasper.apache-el-8.0.23.M1.jar
org.eclipse.jetty.apache-jsp-9.3.4-SNAPSHOT.jar

Some of the classes in org.mortbay.jasper.apache-jsp wouldn't strictly be needed as they are the machinery to on-the-fly compilation, but I think they're a relatively small number.

2/You've got the pull request for the fix for this one already #69

**Comment by janbartel** [comment]
@ludoch Where are we at with this one? Is providing the jsp-related jars inside the webapp when the jsps have been precompiled something that we want to pursue?

**Comment by ludoch** [comment]
We want to support for the GAE standard product, and we are ver close
internally.
The staging process (done in a maven plugin) can now compile the jsps with
Jetty9 and the latest Jasper from tomcat), and the process then adds in
web-inf/lib a few jars needed by the runtime.
I encourage you to play with it (without deploying, which is now allowed
yet externally) with this Maven plugin configuration:
      <plugin>
        <groupId>com.google.appengine</groupId>
        <artifactId>appengine-maven-plugin</artifactId>
        <version>1.9.38</version>
        <configuration>
          <appId>YOURAPPID</appId>
          <versionyourversion</version>
          <retainUploadDir>true></retainUploadDir>
          <additionalParams>
            <param>-R</param>
            <param>--runtime=java8g</param>
            <param>--use_java8</param>
          </additionalParams>
        </configuration>
      </plugin>

the key part here is the retain directory that will contain a modified
exploded war you can look into.

mvn appengine:update
will fail in deployment, but you can see the exploded war content...
Assuming you have 1 or more JSPs in your project.

Maybe you can validate this content? If yes, then we are good to close this
issue.
